### PR TITLE
added aal atlas download and updated reference.rst

### PR DESCRIPTION
**Pull request by MartinPerez**
@GaelVaroquaux I added the fetch_atlas_aal function following example from the other atlas downloaded as requested in #706 

For now I implemented the download of the SPM12 version, any other SPM version would require to hardcode reference to a lot matlab files. I dont know if this would actually be useful for the users, but can implement it if you want. What do you think?

### CODE REVIEW

**Comment by MartinPerez** [comment]
added test

File: nilearn/datasets/atlas.py [hunk 1]
```diff
@@ -576,3 +576,73 @@ def fetch_yeo_2011_atlas(data_dir=None, url=None, resume=True, verbose=1):
         url=url,
         resume=resume,
         verbose=verbose)
+
+
+def fetch_atlas_aal(data_dir=None, url=None, resume=True, verbose=1):
+    """Download and return file names for the AAL for SPM 12 parcellation.
```
Comment (AlexandreAbraham) [inline nilearn/datasets/atlas.py]: It does not download file names ;). Can you just put `Downloads and returns the AAL template for SPM 12.`

File: nilearn/datasets/atlas.py [hunk 2]
```diff
@@ -576,3 +576,73 @@ def fetch_yeo_2011_atlas(data_dir=None, url=None, resume=True, verbose=1):
...
+
+    The provided images are based on the spatially normalized single-subject
```
Comment (AlexandreAbraham) [inline nilearn/datasets/atlas.py]: From that sentence I have the impression that AAL was computed on the MNI T1 template. You can simply say `This atlas has been registered to MNI space` or something like that. Don't we say that for other atlases? If no, then we should.

File: nilearn/datasets/atlas.py [hunk 3]
```diff
@@ -576,3 +576,73 @@ def fetch_yeo_2011_atlas(data_dir=None, url=None, resume=True, verbose=1):
...
+        - "labels": labels id in an xml file
+
+    Notes
+    -----
+    For more information on this dataset's structure, see
+    http://www.gin.cnrs.fr/AAL-217?lang=en
+
+    N. Tzourio-Mazoyer, B. Landeau, D. Papathanassiou, F. Crivello,
+    O. Etard, N. Delcroix, B. Mazoyer, and M. Joliot.
+    NeuroImage 2002. 15 :273-28
```
Comment (AlexandreAbraham) [inline nilearn/datasets/atlas.py]: The title is missing.

File: nilearn/datasets/atlas.py (hunk 2 above)
Comment (AlexandreAbraham) [inline nilearn/datasets/atlas.py]: Deleted since it's the case.

File: nilearn/datasets/atlas.py [hunk 4]
```diff
@@ -576,3 +576,73 @@ def fetch_yeo_2011_atlas(data_dir=None, url=None, resume=True, verbose=1):
         url=url,
         resume=resume,
         verbose=verbose)
+
+
+def fetch_atlas_aal(data_dir=None, url=None, resume=True, verbose=1):
```
Comment (AlexandreAbraham) [inline nilearn/datasets/atlas.py]: So, apparently, the ICBM template is in `datasets.struct` and not here. So, should this one be here or struct? I believe it belongs here.

File: nilearn/datasets/atlas.py [hunk 5]
```diff
@@ -576,3 +576,73 @@ def fetch_yeo_2011_atlas(data_dir=None, url=None, resume=True, verbose=1):
...
+
+    verbose: int
+        verbosity level (0 means no message).
+
+    Returns
+    -------
+    data: sklearn.datasets.base.Bunch
+        dictionary-like object, keys are:
+
+        - "anat": anatomy image.
```
Comment (AlexandreAbraham) [inline nilearn/datasets/atlas.py]: According to nilearn semantics, this one should be called `labels` and the XML file `labels` too. I think we have a problem ;).

File: nilearn/datasets/atlas.py [hunk 6]
```diff
@@ -576,3 +576,73 @@ def fetch_yeo_2011_atlas(data_dir=None, url=None, resume=True, verbose=1):
...
+    For more information on this dataset's structure, see
+    http://www.gin.cnrs.fr/AAL-217?lang=en
+
+    N. Tzourio-Mazoyer, B. Landeau, D. Papathanassiou, F. Crivello,
+    O. Etard, N. Delcroix, B. Mazoyer, and M. Joliot.
+    NeuroImage 2002. 15 :273-28
+
+    Licence: unknown.
+    """
+    spm_version = 12
```
Comment (AlexandreAbraham) [inline nilearn/datasets/atlas.py]: Before going further on this point: is there actually a difference between AAL in SPM8 and in SPM12?

File: nilearn/datasets/atlas.py [hunk 7]
```diff
@@ -576,3 +576,73 @@ def fetch_yeo_2011_atlas(data_dir=None, url=None, resume=True, verbose=1):
...
+    if url is None:
+        baseurl = "http://www.gin.cnrs.fr/AAL/aal_for_SPM%i.tar.gz"
+        url = baseurl % spm_version
+    opts = {'uncompress': True}
+
+    dataset_name = "aal"
+    # keys and basenames would need to be handled for each spm_version
+    # for now spm_version 12 is hardcoded.
```
Comment (AlexandreAbraham) [inline nilearn/datasets/atlas.py]: My gut feeling about that is that a name is rarely too specific. So `aal_spm_12` would maybe be the best option. If we want to add another atlas computed with the AAL methods later, we won't have collisions.

File: nilearn/datasets/atlas.py [hunk 8]
```diff
@@ -576,3 +576,73 @@ def fetch_yeo_2011_atlas(data_dir=None, url=None, resume=True, verbose=1):
...
+    keys = ("anat",
```
Comment (AlexandreAbraham) [inline nilearn/datasets/atlas.py]: You don't need newlines here.

File: nilearn/datasets/atlas.py [hunk 9]
```diff
@@ -576,3 +576,73 @@ def fetch_yeo_2011_atlas(data_dir=None, url=None, resume=True, verbose=1):
...
+            "labels")
+    basenames = (
+        "AAL.nii",
+        "AAL.xml")
+
+    filenames = [(os.path.join("aal_for_SPM%i" % spm_version, f), url, opts)
```
Comment (AlexandreAbraham) [inline nilearn/datasets/atlas.py]: Well, you totally got the `_fetch_files` logic ;).

File: nilearn/datasets/atlas.py [hunk 10]
```diff
@@ -576,3 +576,73 @@ def fetch_yeo_2011_atlas(data_dir=None, url=None, resume=True, verbose=1):
...
+
+    data_dir = _get_dataset_dir(dataset_name, data_dir=data_dir,
+                                verbose=verbose)
+    sub_files = _fetch_files(data_dir, filenames, resume=resume,
+                             verbose=verbose)
+
+    fdescr = _get_dataset_descr(dataset_name)
+
+    params = dict([('description', fdescr)] + list(zip(keys, sub_files)))
+    return Bunch(**params)
```
Comment (AlexandreAbraham) [inline nilearn/datasets/atlas.py]: Putting here the explicit names would maybe be clearer. Also, we should load the XML file for the user. Most of them won't be able to do it.

File: nilearn/datasets/atlas.py (hunk 5 above)
Comment (MartinPerez) [inline nilearn/datasets/atlas.py]: I guess we could say labels and labels_description/labels_ids then?

File: nilearn/datasets/atlas.py (hunk 3 above)
Comment (MartinPerez) [inline nilearn/datasets/atlas.py]: my bad, its this:
Automated Anatomical Labeling of Activations in SPM Using a Macroscopic Anatomical Parcellation of the MNI MRI Single-Subject Brain.

I will add it

File: nilearn/datasets/atlas.py (hunk 6 above)
Comment (MartinPerez) [inline nilearn/datasets/atlas.py]: yes they are quite different, i downloaded all the versions to test, spm12 is the first to have only a nii file and xml file, all the previous ones have a crazy amount of files, i think they have one file per label.

File: nilearn/datasets/atlas.py (hunk 7 above)
Comment (MartinPerez) [inline nilearn/datasets/atlas.py]: I agree with you

File: nilearn/datasets/atlas.py (hunk 10 above)
Comment (MartinPerez) [inline nilearn/datasets/atlas.py]: what do you mean by loading the XML, putting it in a dictionary?

File: nilearn/datasets/atlas.py (hunk 5 above)
Comment (AlexandreAbraham) [inline nilearn/datasets/atlas.py]: We can change the name but in that case, we have to change `harvard_oxford` too to be homogeneous. No good naming comes to my mind we I can summon our naming specialist @banilo ;).

File: nilearn/datasets/atlas.py (hunk 6 above)
Comment (AlexandreAbraham) [inline nilearn/datasets/atlas.py]: Let's stick with the last version.

File: nilearn/datasets/atlas.py (hunk 10 above)
Comment (AlexandreAbraham) [inline nilearn/datasets/atlas.py]: I mean actually loading the content of the XML file and provide a dictionnary associating the label id with the region name.

**Comment by MartinPerez** [comment]
@AlexandreAbraham I fixed all the comments, and about the naming I finally changed anat for maps, which i think follow the convention behind the other atlases.

**Comment by MartinPerez** [comment]
I realized that when it comes to what to return there is no standard across atlases, some return directly a nibable.Niftiimage while others return the file name, with respect to labels, some just return a list while I am returning a dictionary... Should this be reported as an issue?

**Comment by AlexandreAbraham** [comment]
> about the naming I finally changed anat for maps

This is misleading: your atlas cannot be trated by NiftiMapsMasker, it should be fed to NiftiLabelsMasker.

> some return directly a nibable.Niftiimage while others return the file name

When a Nifti1Image is returned, it is probably that some processing of the atlas was required. However, this doesn't matter as both are niimgs and can be processed by all nilearn functions.

> some just return a list while I am returning a dictionary

Well, a list can be seen as an "int to string" dictionary :P. I think that we should return dictionaries everywhere.

> Should this be reported as an issue?

Yes! We definitely need to harmonize the naming conventions and return types. However, I have no time to do it. Let's merge your PR and then solve the problem of the naming.

**Comment by MartinPerez** [comment]
sorry i didnt understand the first comment, so what should I name it finally?

**Comment by AlexandreAbraham** [comment]
I don't know.

**Comment by bthirion** [comment]
Regarding the naming: maybe label_masks or multi_mask ?

File: nilearn/datasets/atlas.py [hunk 11]
```diff
@@ -1,7 +1,8 @@
 """
-Downloading NeuroImaging datasets: altas datasets
+Downloading NeuroImaging datasets: atlas datasets
 """
 import os
+import xml.etree.ElementTree as ET
```
Comment (bthirion) [inline nilearn/datasets/atlas.py]: Maybe more simply: import xml.etree.ElementTree
as this is used only once.

**Comment by bthirion** [comment]
No other comment atm.

**Comment by MartinPerez** [comment]
I changed the import and the name to multi_mask then

**Comment by AlexandreAbraham** [comment]
For me, it's even worse. People will try to stick it into the MultiNiftiMasker and wonder why it doesn't work.

**Comment by MartinPerez** [comment]
@AlexandreAbraham  What if we use then something like "regions" and clarify in the documentation what people are supposed to do with the file?

**Comment by AlexandreAbraham** [comment]
Yes, any naming like "atlas", "regions" or even "label_masks" is fine with me. I just don't want it to collide with other maskers name. We'll change it afterward when we'll agree on the right vocabulary.

**Comment by MartinPerez** [comment]
Ok the name is 'regions' now

**Comment by AlexandreAbraham** [comment]
You have failing tests.

**Comment by AlexandreAbraham** [comment]
Seems like your failing test is related to the test, which is normal because ot tries to load an empty XML file. Take a look at the harvard oxford fetch function: we create a dummy XML file to avoid this kind of crash.

**Comment by MartinPerez** [comment]
thank you for the tip @AlexandreAbraham, I think I fixed the test now following your advice.

**Comment by lesteve** [comment]
It's getting there but you still have an error for Python 2.6 due to the fact that the `.iter` method is new in python 2.7. It looks like `.getiterator` although deprecated works for all the python versions we care about.

**Comment by MartinPerez** [comment]
@lesteve thanks for the advice. The checks seem to be fine finally.

File: nilearn/datasets/atlas.py [hunk 12]
```diff
@@ -576,3 +577,82 @@ def fetch_yeo_2011_atlas(data_dir=None, url=None, resume=True, verbose=1):
...
+        baseurl = "http://www.gin.cnrs.fr/AAL/aal_for_SPM%i.tar.gz"
+        url = baseurl % spm_version
+    opts = {'uncompress': True}
+
+    dataset_name = "aal_spm_12"
+    # keys and basenames would need to be handled for each spm_version
+    # for now spm_version 12 is hardcoded.
+    keys = ("regions", "labels")
+    basenames = (
+        "AAL.nii", "AAL.xml")
```
Comment (lesteve) [inline nilearn/datasets/atlas.py]: Why not put that on a single line?

File: nilearn/datasets/atlas.py [hunk 13]
```diff
@@ -576,3 +577,82 @@ def fetch_yeo_2011_atlas(data_dir=None, url=None, resume=True, verbose=1):
...
+
+    # We return the labels contained in the xml file as a dictionary
+    xml_tree = xml.etree.ElementTree.parse(labels_file)
+    root = xml_tree.getroot()
+    labels_dict = {}
+    for label in root.getiterator('label'):
+        labels_dict[label.find('index').text] = label.find('name').text
+
+    params = dict([('description', fdescr)] + [(keys[0], atlas_img)] +
+                  [(keys[1], labels_dict)])
```
Comment (lesteve) [inline nilearn/datasets/atlas.py]: It looks like you are only using `keys` once so I would remove the `keys` definition above and do:
```python
params = {'descr': fdescr, 'regions': atlas_img, 'labels': labels_dict}
```

File: nilearn/datasets/tests/test_atlas.py [hunk 14]
```diff
@@ -266,3 +266,19 @@ def test_fetch_atlas_yeo_2011():
...
     assert_true(isinstance(dataset.thin_7, _basestring))
     assert_equal(len(mock_url_request.urls), 1)
+
+
+@with_setup(setup_mock, teardown_mock)
+@with_setup(setup_tmpdata, teardown_tmpdata)
+def test_fetch_atlas_aal_spm_12():
+    ho_dir = os.path.join(tmpdir, 'aal_spm_12', 'aal_for_SPM12')
+    os.makedirs(ho_dir)
+    dummy = open(os.path.join(ho_dir, 'AAL.xml'), 'w')
```
Comment (lesteve) [inline nilearn/datasets/tests/test_atlas.py]: you can use a context manager here:

```python
with open(os.path.join(ho_dir, 'AAL.xml'), 'w') as xml_file:
    xml_file.write(...)
```

and the file will be closed at the exit of the block.

File: nilearn/datasets/atlas.py [hunk 15]
```diff
@@ -576,3 +577,82 @@ def fetch_yeo_2011_atlas(data_dir=None, url=None, resume=True, verbose=1):
...
+
+    filenames = [(os.path.join("aal_for_SPM%i" % spm_version, f), url, opts)
+                 for f in basenames]
+
+    data_dir = _get_dataset_dir(dataset_name, data_dir=data_dir,
+                                verbose=verbose)
+    atlas_img, labels_file = _fetch_files(data_dir, filenames, resume=resume,
+                                          verbose=verbose)
+
+    fdescr = _get_dataset_descr(dataset_name)
```
Comment (lesteve) [inline nilearn/datasets/atlas.py]: Would it make sense to add a description of the dataset in nilearn/datasets/description? All you need to do is add aal_spm_12.rst there I think.

**Comment by lesteve** [comment]
Do we want to have fetch_atlas_aal_spm_12 exposed in `nilearn/datasets/__init__.py`?

File: nilearn/datasets/atlas.py (hunk 15 above)
Comment (MartinPerez) [inline nilearn/datasets/atlas.py]: Adding the atlas is @GaelVaroquaux  idea so im not sure what extra documentation is useful or desirable. I will add it you want.

**Comment by MartinPerez** [comment]
@lesteve I made the changes you pointed out, except for the description and exposing the function in __init__. I await your confirmation on what to do about that.

**Comment by AlexandreAbraham** [comment]
Yes we want a description if possible (not mandatory, you can skip it or make it short), and yes it should be exposed like the others :).

**Comment by MartinPerez** [comment]
ok exposed it and added the description file, but im not sure is snough to add aal_spm_12.rst to descriptions folder?

**Comment by AlexandreAbraham** [comment]
Yes it is. This line `fdescr = _get_dataset_descr(dataset_name)` does all the magic of loading it for you ;).

**Comment by MartinPerez** [comment]
but what is the convention? yeo for example should be yeo_2011.rst but instead it is just yeo.rst ... well happy its done then :)

**Comment by AlexandreAbraham** [comment]
> but what is the convention? yeo for example should be yeo_2011.rst but instead it is just yeo.rst ... well happy its done then :)

Seems like an error to me...

**Comment by lesteve** [comment]
> Seems like an error to me...

Good catch, I just fixed it in master.

File: nilearn/datasets/__init__.py [hunk 16]
```diff
@@ -19,4 +19,5 @@
            'fetch_harvard_oxford', 'fetch_atlas_msdl', 'fetch_msdl_atlas',
            'fetch_atlas_power_2011', 'fetch_atlas_smith_2009',
            'fetch_smith_2009', 'fetch_atlas_yeo_2011',
-           'fetch_yeo_2011_atlas', 'fetch_mixed_gambles']
+           'fetch_yeo_2011_atlas', 'fetch_mixed_gambles',
+           'fetch_atlas_aal_spm_12']
```
Comment (lesteve) [inline nilearn/datasets/__init__.py]: About exposing fetch_atlas_aal_spm12, you want to add fetch_atlas_aal_smp12 in the import from .atlas a few lines above so that you can do:

```python

from nilearn import datasets
datasets.fetch_atlas_aal_smp12()
```

For the record `__all___` is used when doing `from nilearn.datasets import *`.

**Comment by MartinPerez** [comment]
I see, thanks @lesteve , done!

**Comment by lesteve** [comment]
LGTM merging, thanks!

**Comment by MartinPerez** [comment]
Nice :). thank you.
